File: frontend/src/api/services/VideoSearchService.ts
import api from '../apiClient'
import type { VideoSortType } from '../interfaces/common'
import type { FollowInfoResponse } from '../interfaces/response'

class VideoSearchService {
  private static path = '/api/video/search'

  // Private 메서드로 API 호출 로직 관리
  private static async _getPaginatedAndSortedVideos(
    keyword: string,
    page: number,
    size: number,
    sort: VideoSortType,
  ): Promise<void> {
    const response = await api.get(this.path, {
      params: {
        keyword,
        page,
        size,
        sort,
      },
    })
    return response.data
  }

  private static async _getAutocompleteSuggestions(keyword: string): Promise<FollowInfoResponse> {
    const response = await api.get(this.path + '/autocomplete', {
      params: {
        keyword,
      },
    })
    return response.data
  }

  // Public 메서드로 외부에서 사용할 수 있도록 제공
  static async getPaginatedAndSortedVideos(
    keyword: string,
    page: number,
    size: number,
    sort: VideoSortType,
  ): Promise<void> {
    return await this._getPaginatedAndSortedVideos(keyword, page, size, sort)
  }

  static getAutocompleteSuggestions(keyword: string): Promise<FollowInfoResponse> {
    return this._getAutocompleteSuggestions(keyword)
  }
}

export default VideoSearchService

File: backend/src/main/java/com/ssafy/ssafit/service/VideoSearchService.java
package com.ssafy.ssafit.service;

import com.ssafy.ssafit.dao.VideoMapper;
import com.ssafy.ssafit.dao.VideoSearchMapper;
import com.ssafy.ssafit.domain.Video;
import com.ssafy.ssafit.domain.VideoSortType;
import com.ssafy.ssafit.dto.response.AutoCompleteVO;
import com.ssafy.ssafit.dto.response.PagedResponseDTO;
import com.ssafy.ssafit.dto.response.VideoCardVO;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;

import static com.ssafy.ssafit.utils.DTOMapper.toPagedResponseDTO;
import static com.ssafy.ssafit.utils.DTOMapper.toVideoCardVO;

@Service
@RequiredArgsConstructor
public class VideoSearchService {

    private final VideoSearchMapper videoSearchMapper;

    public PagedResponseDTO<VideoCardVO> getPaginatedAndSortedVideos(String keyword, int page, int size, String sort) {
        int offset = (page - 1) * size;

        // 데이터베이스에서 검색된 비디오 목록 조회
        List<Video> videos = videoSearchMapper.searchVideos(keyword, offset, size, sort);

        System.out.println(videos);

        // 총 검색 결과 수 조회
        int totalCount = videoSearchMapper.countVideos(keyword);

        // 페이지네이션 정보를 포함한 응답 객체 생성
        return toPagedResponseDTO(page, size, totalCount, toVideoCardVO(videos));
    }

    public List<AutoCompleteVO> getAutocompleteSuggestions(String keyword) {
        List<AutoCompleteVO> autoCompleteSuggestions = videoSearchMapper.getAutocompleteSuggestions(keyword); // 최대 5개 반환
        System.out.println("auto: " + autoCompleteSuggestions);
        return autoCompleteSuggestions;
    }
}